src/storage/src/source/mysql/replication.rs

```diff
@@ -0,0 +1,405 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::{BTreeMap, BTreeSet};
+use std::convert::Infallible;
+use std::pin::pin;
+use std::rc::Rc;
+use std::str::FromStr;
+use std::time::Duration;
+
+use bytes::Bytes;
+use differential_dataflow::{AsCollection, Collection};
+use futures::StreamExt;
+use mysql_async::prelude::Queryable;
+use mysql_async::{BinlogStream, BinlogStreamRequest, Conn, GnoInterval, Sid};
+use timely::dataflow::channels::pact::Exchange;
+use timely::dataflow::operators::{Concat, Map};
+use timely::dataflow::{Scope, Stream};
+use timely::progress::Antichain;
+use timely::progress::Timestamp;
+use tracing::trace;
+use uuid::Uuid;
+
+use mz_mysql_util::GtidSet;
+use mz_mysql_util::{
+    ensure_full_row_binlog_format, ensure_gtid_consistency, ensure_replication_commit_order,
+    MySqlTableDesc,
+};
+use mz_ore::cast::CastFrom;
+use mz_ore::result::ResultExt;
+use mz_repr::{Diff, GlobalId, Row};
+use mz_sql_parser::ast::UnresolvedItemName;
+use mz_storage_types::sources::MySqlSourceConnection;
+use mz_timely_util::builder_async::{
+    Event as AsyncEvent, OperatorBuilder as AsyncOperatorBuilder, PressOnDropButton,
+};
+
+use crate::source::types::SourceReaderError;
+use crate::source::RawSourceCreationConfig;
+
+use super::{
+    pack_mysql_row, table_name, DefiniteError, ReplicationError, RewindRequest, TransactionId,
+    TransientError,
+};
+
+// Used as a partition id to determine if the worker is responsible for reading from the
+// MySQL replication stream
+static REPL_READER: &str = "reader";
+
+// A constant arbitrary offset to add to the source-id to produce a deterministic server-id for
+// identifying Materialize as a replica on the upstream MySQL server.
+// TODO(roshan): Add documentation for this
+static REPLICATION_SERVER_ID_OFFSET: u32 = 524000;
+
+/// Renders the replication dataflow. See the module documentation for more information.
+pub(crate) fn render<G: Scope<Timestamp = TransactionId>>(
+    scope: G,
+    config: RawSourceCreationConfig,
+    connection: MySqlSourceConnection,
+    subsource_resume_uppers: BTreeMap<GlobalId, Antichain<TransactionId>>,
+    table_info: BTreeMap<UnresolvedItemName, (usize, MySqlTableDesc)>,
+    rewind_stream: &Stream<G, RewindRequest>,
+    _committed_uppers: impl futures::Stream<Item = Antichain<TransactionId>> + 'static,
```

Comment:
does mysql even have the same concept of pg, where you tell it up to what lsn (gtid in this case) it can compact the data to?

Reply:
Nope, there is no reporting mechanism that a mysql replica uses to tell the mysql source what it has committed. The source does [keep track](https://dev.mysql.com/doc/refman/8.0/en/source-thread-states.html) of what it has sent to actively connected replicas, but this is not persisted like a postgres replication-slot is (when the replica disconnects, the source forgets about it).

The automatic binary log purging is just done based on a configuration value on the mysql source:
https://dev.mysql.com/doc/refman/8.0/en/purge-binary-logs.html
> Binary log files are automatically removed after the server's binary log expiration period. Removal of the files can take place at startup and when the binary log is flushed. The default binary log expiration period is 30 days. You can specify an alternative expiration period using the [binlog_expire_logs_seconds](https://dev.mysql.com/doc/refman/8.0/en/replication-options-binary-log.html#sysvar_binlog_expire_logs_seconds) system variable. If you are using replication, you should specify an expiration period that is no lower than the maximum amount of time your replicas might lag behind the source.